File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        balanced, depth = self.dfs(root)

        return balanced

    def dfs(self, node):
        if node == None:
            return [True, 0]

        leftBalanced, leftDepth = self.dfs(node.left)
        rightBalanced, rightDepth = self.dfs(node.right)

        currentNodeBalanced = (
            leftBalanced and rightBalanced and abs(leftDepth - rightDepth) <= 1
        )

        return [currentNodeBalanced, max(leftDepth, rightDepth) + 1]

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function (root) {
    const dfs = function (node) {
        if (!node) return [true, 0];

        const [leftBalanced, leftDepth] = dfs(node.left);
        const [rightBalanced, rightDepth] = dfs(node.right);

        let currentNodeBalanced = (Math.abs(leftDepth - rightDepth) <= 1 && leftBalanced && rightBalanced);

        return [currentNodeBalanced, Math.max(leftDepth, rightDepth) + 1 ];
    }

    const [ balanced ] = dfs(root);
    
    return balanced;
};